src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,12 +394,156 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            IEdmTypeReference elementType = edmPropertyType.ElementType();
+            ODataResourceSetWrapper resourceSetWrapper = new ODataResourceSetWrapper(resourceSet);
+            foreach (ODataEntityReferenceLinkWrapper refLinkWrapper in refLinks)
+            {
+                ODataResourceWrapper resourceWrapper = CreateResourceWrapper(elementType, refLinkWrapper, readContext);
+                resourceSetWrapper.Resources.Add(resourceWrapper);
+            }
+
+            return resourceSetWrapper;
+        }
+
+        private ODataResourceWrapper CreateResourceWrapper(IEdmTypeReference edmPropertyType, ODataEntityReferenceLinkWrapper refLink, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(readContext != null);
+
+            ODataResource resource = new ODataResource
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            Uri serviceRootUri = null;
+            string serviceRoot = "";
+            DefaultODataPathHandler pathHandler = new DefaultODataPathHandler();
+
+            IWebApiRequestMessage internalRequest = readContext.InternalRequest;
+            IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+            serviceRoot = readContext.InternalUrlHelper.CreateODataLink(
+                readContext.InternalRequest.Context.RouteName,
+                readContext.InternalRequest.PathHandler,
+                new List<ODataPathSegment>());
+
+            serviceRootUri = new Uri(serviceRoot, UriKind.Absolute);
+
+            var request = readContext.Request;
+            IEdmModel model = readContext.Model;
+
+            ODataPath odataPath = pathHandler.Parse(serviceRoot, refLink.EntityReferenceLink.Url.OriginalString,readContext.InternalRequest.RequestContainer);
+            IList<KeySegment> keySegments = odataPath.Segments.OfType<KeySegment>().ToList();
+            if (keySegments == null)
+            {
+                return null;
+            }
+
+            foreach (KeySegment keySegment in keySegments)
+            {
+                resource.Properties = keySegment.Keys.Select(k => new ODataProperty
+                {
+                    Name = k.Key,
+                    Value = k.Value
+                });
+            }
+
+            if (refLink.EntityReferenceLink.InstanceAnnotations != null)
+            {
+                foreach (ODataInstanceAnnotation instanceAnnotation in refLink.EntityReferenceLink.InstanceAnnotations)
+                {
+                    resource.InstanceAnnotations.Add(instanceAnnotation);
+                };
+            }
+            
+            return new ODataResourceWrapper(resource);
+        }
+
+        private ODataResourceWrapper UpdateResourceWrapper(IEdmTypeReference edmPropertyType, Uri resourceId, ODataDeserializerContext readContext)
```

Comment:
A fair amount of logic in this method is a mirror of the logic in `CreateResourceWrapper` method. You should refactor the shared logic into a method that both methods can call